### DISCUSSION TITLE
Does Nextauth rotate the session token for email auth?

### DISCUSSION THREAD

**Discussion by anguskeatinge** [nextauthjs/next-auth#3178] [Help]
I'm running into some bugs (I think) with the session not being persisted properly and I want to handle this myself.

```
const options = {
    ...
    providers: [
        EmailProvider({
            ...
        }),
    ],
    adapter: MongoDBAdapter({ db }),
    session: {
        jwt: false,
        maxAge: 30 * 24 * 60 * 60, // 30 days
        updateAge: 24 * 60 * 60, // 1 day
    },
   ...
}
```

**Comment by balazsorban44** [comment]
Will need a reproduction.

**Reply by anguskeatinge** [reply]
I'm going to worry about the bugs later, for now I'm just asking the question.

**Reply by balazsorban44** [reply]
Session rotation is not provider-specific, so yes. https://github.com/nextauthjs/next-auth/blob/beta/src/core/routes/session.ts

**Reply by anguskeatinge** [reply]
I can see that next-auth is updating the expiry on the sessions, but the session token itself seems to be permanent.

**Reply by balazsorban44** [reply]
https://github.com/nextauthjs/next-auth/blob/043b252940e262361dbe62816793ccc890ca71f2/src/core/routes/session.ts#L146-L154

**Reply by anguskeatinge** [reply]
Yep, that's what I'm referring to. You update the expiry, but the sessionToken itself isn't rotated.

https://github.com/nextauthjs/next-auth/blob/043b252940e262361dbe62816793ccc890ca71f2/src/core/routes/session.ts#L123-L125

**Reply by anguskeatinge** [reply]
This way, a session token could last indefinitely.

**Reply by anguskeatinge** [reply]
It's not a massive issue, but if someone managed to brute force the uuid (`randomBytes(32).toString('hex')`) of a super user (or if it was exposed some other way) then they would potentially have permanent access to that account.

**Reply by anguskeatinge** [reply]
It would probably make sense to rotate the sessionToken and probably sign it as well (e.g. `jwt.sign()`). 
☝️ I'm doing this now - it doesn't seem like much extra work for a bunch more security.

**Reply by balazsorban44** [reply]
Thanks. After checking https://cheatsheetseries.owasp.org/cheatsheets/Session_Management_Cheat_Sheet.html#renew-the-session-id-after-any-privilege-level-change

I did not find a recommendation for token rotation on every session check.

I did understand that we could probably generate a harder to guess session token though

https://github.com/nextauthjs/next-auth/blob/043b252940e262361dbe62816793ccc890ca71f2/src/core/lib/callback-handler.ts#L224

any suggestions?

**Reply by anguskeatinge** [reply]
Thanks for the resource 👍 
Let me scout about for some resources to back up my approach, I'll get back to you on that.

Fwiw I'm only rotating the sessionToken when it we update the expiry, i.e. daily. 
And then I am setting the expiry of the old session (in the db) to 15 mins away so concurrent requests with old session tokens don't fail. (I recognise that I'll need a job to clean up old sessions to prevent bloat)

Hmm, well the `randomUUID` is kinda meaningless atm because the latest runtime for aws Lambdas is 14.x
This guy reckons we should use the `uuid` package.
https://gist.github.com/joepie91/7105003c3b26e65efcea63f3db82dfba
https://www.npmjs.com/package/uuid

**Reply by anguskeatinge** [reply]
Haha, after reading that joepie gist in more detail, it looks like `randomBytes(32).toString("hex")` is probably acceptable because we aren't modding or transforming the output.

**Reply by anguskeatinge** [reply]
So it turns out updating the session on the backend is fraught because dealing with concurrent requests is a hassle.

I guess the only way to deal with that is to make sure the SessionProvider is the only thing that calls `/api/auth/session`. 
Then the only potential for concurrency is when you're opening up multiple tabs at the exact same time...

Haha, look, I'm just going to hack away at this and I'll let you know how I go when the dust has settled. 😄
